draft-ietf-gnap-resource-servers.md

```diff
@@ -0,0 +1,502 @@
+: A list of token formats supported by this AS.
+
+resource_registration_endpoint:
+: The URL of the endpoint offering resource registration. {{rs-register-resource-handle}}
+
+grant_endpoint:
+: The grant endpoint of the GNAP AS.
+
+## Protecting RS requests to the AS {#authentication}
+
+Where noted, the RS protects its calls to the AS using any of the signature
+methods defined by GNAP. This signing method MUST cover all of the appropriate
+
+The AS MAY require an RS to pre-register its keys
+or could alternatively allow calls from arbitrary keys, in a trust-on-first-use
+model. The RS MAY present its keys by reference or by value in
+the same fashion as a client instance calling the AS in the core protocol
+of GNAP {{I-D.ietf-gnap-core-protocol}}.
+
+## Token Introspection {#introspection}
+
+The AS issues access tokens representing a set of delegated access rights
+to be used at one or more RSs. The AS can offer an introspection service
+to allow an RS to validate that a given access token:
+
+ - has been issued by the AS
+ - has not expired
+ - has not been revoked
+ - is appropriate for the RS identified in the call
+
+When the RS receives an access token, it can call the introspection
+endpoint at the AS to get token information. 
+\[\[ [See issue #115](https://github.com/ietf-wg-gnap/gnap-core-protocol/issues/115) \]\]
+
+~~~
++--------+       +------+       +------+
+| Client |--(1)->|  RS  |       |  AS  |
+|Instance|       |      |--(2)->|      |
+|        |       |      |       |      |
+|        |       |      |<-(3)--|      |
+|        |       |      |       +------+
+|        |<-(4)--|      |               
++--------+       +------+               
+~~~
+
+1. The client instance calls the RS with its access token.
+
+2. The RS introspects the access token value at the AS.
+    The RS signs the request with its own key (not the client instance's
+    key or the token's key).
+
+3. The AS validates the access token value and the client instance's request
+    and returns the introspection response for the token.
+
+4. The RS fulfills the request from the client instance.
+
+The RS signs the request with its own key and sends the access
+token as the body of the request.
+
+~~~
+POST /introspect HTTP/1.1
+Host: server.example.com
+Content-Type: application/json
+Detached-JWS: ejy0...
+
+{
+    "access_token": "OS9M2PMHKUR64TB8N6BW7OZB8CDFONP219RP1LT0",
+    "proof": "httpsig",
+    "resource_server": "7C7C4AZ9KHRS6X63AJAO"
+}
+~~~
+
+
+
+The AS responds with a data structure describing the token's
+current state and any information the RS would need to validate the
+token's presentation, such as its intended proofing mechanism and key
+material. The response MAY include any fields defined in an access
+token response.
+
+~~~
+HTTP/1.1 200 OK
+Content-Type: application/json
+Cache-Control: no-store
+
+{
+    "active": true,
+    "access": [
+        "dolphin-metadata", "some other thing"
+    ],
+    "key": {
```

Comment:
What's about the rotation of keys? Shouldn't it be a list of keys? Without that, we won't have a smooth transition when the client decides to rotate its key.

I still suggest requiring the JWK thumbprint as a value for `kid` in `jwk` as protection against misuse.

Reply:
In this design, if the key is rotated then whatever the "current" key is will be returned here, replacing the "old" key. I'm not sure I see the value in the overhead that multiple keys would have here just to address this one particular corner. To me it feels like having multiple keys tied to a single access token at any time would be problematic and lead to potential security gaps. Introspection already has to deal with cache inconsistency problems, so an RS is going to need to be robust against something like that anyway. This kind of situation can be talked about in some kind of implementation guide section, perhaps.

Reply:
Interestingly... so, your point is that an RS should be robust to handle this. I get it. Although we may still have & deal with an extremely chatty application and there is an unavoidable delay between the decision to rotate its key by the client and what both AS and RS known.

Reply:
I still like to hear your thoughts on

`I still suggest requiring the JWK thumbprint as a value for kid in jwk as protection against misuse.`

Reply:
Regarding `with cache inconsistency problems`, to be completely pedantic, the response, as well as the spec, includes `Cache-Control: no-store`. So, it's forbidden. You may allow omitting it for the introspection endpoint's response.

Reply:
HTTP Cache and functional application cache are totally separate from each other. You can disable an HTTP cache but still have an application save the results and simply not make another HTTP call again.

Reply:
JWK identifiers are varied and can have meaning inside of an environment (such as date/timestamps, service identifiers, etc), requiring a specific format will limit the utility of the protocol unnecessarily as it adds no additional benefits.